### PR TITLE
Feat 514: add comment value score

### LINKED ISSUE #514: Add comment value / relevance to Comment Moderation experiment
## What problem does this address?

The new [Comment Moderation experiment](https://github.com/WordPress/ai/pull/155) introduced sentiment analysis and toxicity for comments, along with extensibility hooks around the analysis pipeline.  During review discussion in #155, additional interest emerged around evaluating the overall value or relevance of a comment, not just whether it is toxic or negative.  This could help identify low-value comments such as:
* spammy or engagement-bait replies
* off-topic responses
* generic “thanks” / "+1" / low-context comments
* comments that do not meaningfully contribute to discussion

In being able to detect those types of comments, it could support functions like:
* moderation workflows
* prioritization for community managers
* surfacing high-quality discussion
* future automation around comment handling

## What is your proposed solution?

Update the Comment Moderation experiment to add a "Comment Value / Relevance" analysis feature that builds on the existing comment analysis infrastructure.  Related discussion in #155 noted that some extensibility hooks already exist around system instructions and response schema, though additional JS extensibility may be needed for the full admin experience.

Potential capabilities:
* Analyze comments for relevance and contribution quality
* Store a normalized “value” or “relevance” score
* Add a dedicated column in the Comments screen and pill on the Activity dashboard widget
* Allow bulk processing similar to sentiment/toxicity analysis
* Optionally flag low-value comments for moderation or review
* Provide filters/hooks for custom scoring logic and thresholds

### PR DESCRIPTION
**Pull request by priyanshuhaldar007**
## What?
Closes https://github.com/WordPress/ai/issues/514

Adds a **Value Score** column to the comment moderation list table. Comments are now analyzed for how relevant and valuable they are to the article they're posted on, in addition to the existing toxicity and sentiment signals.

## Why?

Toxicity and sentiment alone don't tell the full story of a comment's quality. A comment can be perfectly polite but still be spam, a generic "+1", or completely off-topic. This PR introduces a value score (0–1) so moderators can quickly identify substantive, on-topic contributions versus low-effort noise — making triage faster and more informed.

## How?

The implementation follows the exact same pattern as the existing `toxicity_score` field end-to-end:

**AI / Prompt layer**
- Updated `system-instruction.php` to instruct the model to return a third field, `value_score`, scored 0–1 with clear band definitions (low/medium/high). The prompt now also passes post context (excerpt → AI summary → trimmed content fallback) so the model can actually assess relevance against the article.
- Added `get_post_context()` to `Comment_Analysis.php` to fetch and prepare that context, with a graceful 650-char truncation fallback on raw content.
- `analyze_comment()` now accepts and passes `$post_id` through to the prompt builder.

**Schema & storage**
- `output_schema()` and `response_schema()` both declare `value_score` as a nullable float (0–1). It's nullable for cases where the post content is unavailable or too sparse to judge relevance.
- `sanitize_analysis_result()` clamps the value to [0, 1] and preserves `null`.
- Results are stored in a new comment meta key `META_VALUE_SCORE` (`_wpai_value_score`) and returned in the ability response payload.

**Comment Moderation UI (PHP)**
- Added `VALUE_SCORE_LOW / MEDIUM / HIGH` constants and `get_value_score_config()` with the same range-bucket shape used by toxicity, so the frontend JS can resolve badges identically.
- New `wpai_value_score` column registered in `add_columns()` and `add_sortable_columns()`.
- `render_value_score_column()` and `render_value_score_badge()` added, mirroring the toxicity equivalents.
- Filter dropdown added for value score levels.
- `handle_sorting_and_filtering()` extended to support `wpai_value_score` ordering via a meta query, same pattern as toxicity sorting.
- `enqueue_assets()` now passes `value_score` label config into `window.aiCommentModerationData.labels`.
- CSS: reused existing green/yellow/red badge classes by appending `ai-badge--high-value`, `ai-badge--medium-value`, and `ai-badge--low-value` to the existing selectors — no new colour definitions needed.

**Frontend JS / TSX**
- `AnalysisResult` type extended with `value_score: number`.
- `Window` declaration extended with the `value_score` labels shape.
- `PendingComment` type extended with `valueScoreBadge: HTMLElement`.
- `getValueScoreDisplay()` helper added alongside `getToxicityDisplay()`, using the same range-bucket lookup.
- `updateBadges()`, `findPendingComments()`, `analyzeComment()` all updated to handle the third badge — detection, processing state, result rendering, and failure state.

### Use of AI Tools

AI assistance: Yes  
Tool(s): Claude  
Model(s): Claude Sonnet 4.6  
Used for: Drafting this PR description from the git diff. All code was written and reviewed by me, with some contributions from the copilot for updating and generating doc blocks

## Testing Instructions

1. Install and activate the plugin with the AI experiment enabled.
2. Create a post with meaningful content and leave a few comments — mix of on-topic, generic ("great post!"), and spam.
3. Navigate to **Comments** in wp-admin.
4. Confirm a **Value** column appears alongside Sentiment and Toxicity.
5. Trigger analysis (or wait for lazy analysis on page load). Verify badges appear with appropriate levels — high-value comments should show 🌟 High, spam/generic should show ✓ Low.
6. Test the **Value Score** filter dropdown — filtering by Low, Medium, or High should correctly narrow the list.
7. Test sorting by clicking the Value column header — ascending and descending should order comments by their score.
8. Confirm the dashboard comment excerpt pills also show the value score badge.

## Screenshots or screencast

| Before | After |
| ------ | ----- |
| Screenshot 2026-06-08 at 2 06 34 PM | Screenshot 2026-06-08 at 2 05 47 PM |

## Changelog Entry

> Added - Value Score column to the comment moderation table, providing a relevance signal (0–1) for each comment based on how substantive and on-topic it is relative to the post it was left on.

<a href="https://playground.wordpress.net?blueprint-url=data:application/json,%7B%22preferredVersions%22%3A%7B%22wp%22%3A%22latest%22%7D%2C%22login%22%3Atrue%2C%22landingPage%22%3A%22%2Fwp-admin%2Fadmin.php%3Fpage%3Dai-wp-admin%22%2C%22steps%22%3A%5B%7B%22step%22%3A%22installPlugin%22%2C%22pluginData%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fgithub.com%2FWordPress%2Fai%2Freleases%2Fdownload%2Fci-artifacts%2Fpr-681-147f3123a1761a4aadab292c967eadf9712fcddc.zip%22%7D%7D%5D%7D" target="_blank" rel="noopener noreferrer">
  Open WordPress Playground Preview
</a>

### CODE REVIEW

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 1]
```diff
@@ -205,16 +214,50 @@ private function response_schema(): array {
...
 	 * @since 0.9.0
 	 *
 	 * @param string $content The comment content.
 	 * @param string $author  The comment author name.
-	 * @return array{toxicity_score: float, sentiment: string}|\WP_Error The analysis result.
+	 * @param string $post_id The ID of the post.
+	 * @return array{toxicity_score: float, sentiment: string, value_score: float|null}|\WP_Error The analysis result.
 	 */
-	private function analyze_comment( string $content, string $author ) {
+	private function analyze_comment( string $content, string $author, string $post_id ) {
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php]: I believe `$post_id` should always be an integer
```suggestion
	 * @param int $post_id The ID of the post.
	 * @return array{toxicity_score: float, sentiment: string, value_score: float|null}|\WP_Error The analysis result.
	 */
	private function analyze_comment( string $content, string $author, int $post_id ) {
```

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 2]
```diff
@@ -223,20 +266,24 @@ private function analyze_comment( string $content, string $author ) {
 		 *
 		 * @since 0.9.0
 		 *
-		 * @param array{toxicity_score: float, sentiment: string}|null $result  Precomputed analysis result.
-		 * @param string                                              $content Comment content.
-		 * @param string                                              $author  Comment author name.
+		 * @param array{toxicity_score: float, sentiment: string, value_score: float|null}|null $result  Precomputed analysis result.
+		 * @param string                            $content Comment content.
+		 * @param string                            $author  Comment author name.
+		 * @param string                            $post_id The ID of the post.
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php]:
```suggestion
		 * @param int                            $post_id The ID of the post.
```

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 3]
```diff
@@ -66,6 +66,12 @@ protected function output_schema(): array {
 					'enum'        => array_keys( Comment_Moderation::get_sentiment_config() ),
 					'description' => esc_html__( 'The sentiment of the comment.', 'ai' ),
 				),
+				'value_score'    => array(
+					'type'        => array( 'number', 'null' ),
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php]: Is there value in having this return `null`? I think I'd just match what we do for toxicity scoring and have a number between 0 and 1

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php [hunk 4]
```diff
@@ -128,6 +135,63 @@ class Comment_Moderation extends Abstract_Feature {
...
+	 * Value score level: high.
+	 *
+	 * @var string
+	 */
+	public const VALUE_SCORE_HIGH = 'high';
+
+	/**
+	 * Gets the configuration for Value_Score levels.
+	 *
+	 * @since 1.0.0
```
Comment (dkotter) [inline includes/Experiments/Comment_Moderation/Comment_Moderation.php]:
```suggestion
	 * @since x.x.x
```

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php [hunk 5]
```diff
@@ -663,6 +764,30 @@ private function render_toxicity_column( int $comment_id, string $status ): void
 		}
 	}
 
+	/**
+	 * Renders the value score column content.
+	 *
+	 * @since 1.0.0
```
Comment (dkotter) [inline includes/Experiments/Comment_Moderation/Comment_Moderation.php]:
```suggestion
	 * @since x.x.x
```

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php [hunk 6]
```diff
@@ -716,6 +841,37 @@ private function render_toxicity_badge( float $score ): void {
 		);
 	}
 
+	/**
+	 * Renders a value score badge.
+	 *
+	 * @since 1.0.0
```
Comment (dkotter) [inline includes/Experiments/Comment_Moderation/Comment_Moderation.php]:
```suggestion
	 * @since x.x.x
```

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php [hunk 7]
```diff
@@ -716,6 +841,37 @@ private function render_toxicity_badge( float $score ): void {
...
+	 *
+	 * @param float $score The value score (0-1).
+	 */
+	private function render_value_score_badge( float $score ): void {
```
Comment (dkotter) [inline includes/Experiments/Comment_Moderation/Comment_Moderation.php:963]: This is practically the same as `render_toxicity_badge`. Ideally we should refactor to avoid duplicate code here

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php [hunk 8]
```diff
@@ -663,6 +764,30 @@ private function render_toxicity_column( int $comment_id, string $status ): void
...
 
+	/**
+	 * Renders the value score column content.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param int    $comment_id The comment ID.
+	 * @param string $status     The analysis status.
+	 */
+	private function render_value_score_column( int $comment_id, string $status ): void {
```
Comment (dkotter) [inline includes/Experiments/Comment_Moderation/Comment_Moderation.php:892]: Similar here, this method is basically the same as `render_toxicity_column` and `render_sentiment_column`. I'd suggest we introduce a new generic method that can be used by all three of these to avoid duplicate code

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 9]
```diff
@@ -205,16 +214,50 @@ private function response_schema(): array {
 		return (array) apply_filters( 'wpai_comment_analysis_response_schema', $schema );
 	}
 
+	/**
+	 * Function to return context from the post for comment analysis.
+	 * @param int $post_id The ID of the post.
+	 * @return string The content of the post.
+	 */
+	private function get_post_context( int $post_id ): ?string {
+		$post = get_post( $post_id );
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php:226]: The post may not exist so we need an early return check here

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 10]
```diff
@@ -205,16 +214,50 @@ private function response_schema(): array {
...
+
+		// 1. Use excerpt if available (human-written, most reliable)
+		$excerpt = trim( $post->post_excerpt );
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php:233]: Any benefit to using the excerpt or summary over just using the full post content?

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 11]
```diff
@@ -205,16 +214,50 @@ private function response_schema(): array {
...
+		// 3. Fall back to trimmed post content (strip tags, normalize whitespace)
+		$content = wp_strip_all_tags( $post->post_content );
+		$content = preg_replace( '/\s+/', ' ', $content );
+		$content = trim( $content );
+
+		if ( empty( $content ) ) {
+			return null;
+		}
+
+		return mb_substr( $content, 0, 650 );
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php:252]: Not sure we need to trim this. I understand it will save tokens but likely gives us worse results

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php [hunk 12]
```diff
@@ -223,20 +266,24 @@ private function analyze_comment( string $content, string $author ) {
...
+		$post_context = $this->get_post_context( absint( $post_id ) );
+
 		$prompt = sprintf(
-			"Comment by %s:\n\"\"\"%s\"\"\"",
+			"Comment by %s:\n\"\"\"%s\"\"\"\nContext:\n\"\"\"%s\"\"\"",
 			$author,
-			$content
+			$content,
+			$post_context
 		);
```
Comment (dkotter) [inline includes/Abilities/Comment_Moderation/Comment_Analysis.php]: Now that we're passing in the content of the post a comment is on, I'd suggest we change how we build this to send to the LLM. You can look at our other abilities to see how they work but generally we put things in XML-like tags

**Review by dkotter — CHANGES_REQUESTED** [review]
Left a number of comments and also seeing quite a few lint failures that need fixed here. In addition, we should look to add both unit tests and update our E2E tests to ensure these changes are covered

**Comment by priyanshuhaldar007** [comment]
I have implemented changes based on the reviews. For the value score generation fallback, I'd like to know your preference on the method, whether to keep this implementation or simply pass the entire post content and not the excerpt or summary.

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php (hunk 12 above)
Comment (priyanshuhaldar007) [reply includes/Abilities/Comment_Moderation/Comment_Analysis.php]: Added in https://github.com/WordPress/ai/pull/681/commits/06b75af49b385ad192fbcf9a7a867e4fd541cbdb

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php (hunk 9 above)
Comment (priyanshuhaldar007) [reply includes/Abilities/Comment_Moderation/Comment_Analysis.php:226]: Added in https://github.com/WordPress/ai/pull/681/commits/f09061b494e23b662c69888acef5a8f07087f122

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php (hunk 8 above)
Comment (priyanshuhaldar007) [reply includes/Experiments/Comment_Moderation/Comment_Moderation.php:892]: Added in https://github.com/WordPress/ai/pull/681/commits/8bb84d751847eaad7fae69c37ccc066a46fbeafe

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php (hunk 7 above)
Comment (priyanshuhaldar007) [reply includes/Experiments/Comment_Moderation/Comment_Moderation.php:963]: Added in https://github.com/WordPress/ai/pull/681/commits/8bb84d751847eaad7fae69c37ccc066a46fbeafe

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php (hunk 5 above)
Comment (priyanshuhaldar007) [reply includes/Experiments/Comment_Moderation/Comment_Moderation.php]: Applied in https://github.com/WordPress/ai/pull/681/commits/a4f196a8b02f9ffaa700e598c19998dd31e00646

File: includes/Experiments/Comment_Moderation/Comment_Moderation.php (hunk 6 above)
Comment (priyanshuhaldar007) [reply includes/Experiments/Comment_Moderation/Comment_Moderation.php]: Applied in https://github.com/WordPress/ai/pull/681/commits/a4f196a8b02f9ffaa700e598c19998dd31e00646

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php (hunk 2 above)
Comment (priyanshuhaldar007) [reply includes/Abilities/Comment_Moderation/Comment_Analysis.php]: Updated in https://github.com/WordPress/ai/pull/681/commits/9beacfd190c2016f38b04c3a97545b61b28a1aad

File: includes/Abilities/Comment_Moderation/Comment_Analysis.php (hunk 3 above)
Comment (priyanshuhaldar007) [reply includes/Abilities/Comment_Moderation/Comment_Analysis.php]: Updated in https://github.com/WordPress/ai/pull/681/commits/9beacfd190c2016f38b04c3a97545b61b28a1aad

**Comment by jeffpaul** [comment]
@priyanshuhaldar007 FYI a merge conflict to resolve to help keep this moving along

**Comment by dkotter** [comment]
@priyanshuhaldar007 

> For the value score generation fallback, I'd like to know your preference on the method, whether to keep this implementation or simply pass the entire post content and not the excerpt or summary.

I would suggest testing both approaches and seeing what results you get. If you get more or less the same results by using the excerpt/summary, I'd say we're fine to keep that as-is to save tokens. But if you get more accurate / better results by passing in the full content, we should switch to that

**Comment by priyanshuhaldar007** [comment]
I set up a testing framework to measure whether using a post's excerpt/summary or its full content actually changes the quality of the generated value scores.

### Test setup

Created three sample posts covering common scenarios:

* A post with a strong excerpt and detailed content (baseline case)
* A post with a generic/minimal excerpt but a detailed body (to highlight missing context)
* A post with no excerpt (to test fallback behavior)

Added 18 comments across these posts to reflect realistic user interactions:

* 6 thoughtful, relevant comments (expected to score high)
* 6 short, generic comments (expected to score low)
* 4 spam or off-topic comments (expected to score low)
* 2 edge-case comments designed to reveal differences between excerpt and full-content analysis

### Testing framework

Built a browser-based test harness that:

* Scores every comment twice—once using the excerpt and once using the full post content
* Compares the resulting value scores from both approaches
* Checks how closely each score matches the expected High/Medium/Low rating
* Calculates overall accuracy for each mode
* Recommends the better approach based on the final results

### Handling API limitations

To avoid quota and rate-limit issues with OpenAI and Gemini Pro, the test runner:

* Waits 500ms between requests
* Processes comments one at a time instead of in parallel
* Handles failed API calls gracefully, making retries straightforward
* Uses a 300-second timeout for slower responses
* Continues running even if individual comment evaluations fail

### Decision criteria

* If the accuracy difference is under 10%, we'll keep using excerpts to save tokens.
* If full-content analysis improves accuracy by 15% or more, we'll switch to passing the entire post.
* The results provide a clear data-backed trade-off between token usage and scoring accuracy.

This approach replaces assumptions with measurable results, making it easier to decide which context provides the best balance of accuracy and efficiency.

### Results
Excerpt-based generation performed almost as well as full-content generation, with accuracy ranging from 80–100% compared to 80–87.5% for full content. This suggests we can reduce token usage without a noticeable drop in generation quality.

A significant number of test cases returned null responses (8–11 per test cycle). However, these were not due to poor model performance. They were caused by external API issues from AI providers, such as rate limits and temporary server errors (e.g., HTTP 502), which prevented responses from being generated. While these failures should be accounted for when evaluating overall system reliability, they do not reflect the quality of the underlying prompts or generation logic.

Excerpt mode offers better token efficiency, while full-content mode occasionally produces more consistent results. However, these findings are based on a relatively small number of tests (around 8–21 comment generations per cycle), so additional testing is needed before making a final recommendation.

I'm continuing to expand the test dataset, refine the prompts, and evaluate both excerpt and full-content generation across a wider range of scenarios. I'm also investigating retry and fallback strategies to handle transient API failures (such as rate limits and 502 errors) so the comparison reflects model performance rather than external provider issues. This should lead to a more reliable evaluation and a more stable content generation pipeline.

**Comment by priyanshuhaldar007** [comment]
@dkotter, shall I upload the browser-end test script and all the necessary details(post content and comments) here so that you can have a look at them on your end as well?

**Comment by dkotter** [comment]
@priyanshuhaldar007 Thanks for the detailed testing, that's super helpful! Based on the results you're seeing, I think we're fine sticking with the excerpt only for now.

I know there's some failing tests here that need fixed up but is there anything else you're tracking that needs done before this is ready for another review?

**Review by dkotter — APPROVED** [review]
